client/extensions/woocommerce/lib/generate-variations/index.js

```diff
@@ -0,0 +1,36 @@
+// http://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript
+const f = ( a, b ) => [].concat( ...a.map( c => b.map( d => [].concat( c, d ) ) ) );
+const cartesian = ( a, b, ...c ) => b ? cartesian( f( a, b ), ...c ) : a;
+
+/**
+ * Generates variation objects based on a product's attributes.
+ *
+ * @param {Object} product Product object.
+ * @return {Array} Array of variation objects.
+ */
+export default function generateVariations( product ) {
+	const { attributes } = product;
+	const variationTypeOptions = [];
+	const variationTypeNames = [];
+	const variationTypes = (
+		attributes &&
+		attributes.filter( attribute => attribute.variation && attribute.name && attribute.options.length > 0 )
+	) || [];
+
+	variationTypes.forEach( function( variationType ) {
+		variationTypeOptions.push( variationType.options );
+		variationTypeNames.push( variationType.name );
+	} );
+
+	const combinations = cartesian( ...variationTypeOptions );
+	return combinations && combinations.map( function( combination ) {
```

Comment:
I find those `variable && variable.something || defaultValue` statements to be quite confusing. I would rather put `if ( !variationTypeOptions.length ) return [];` a few lines earlier instead.